use a functor for projection (#385)
JuliaDiff/ChainRulesCore.jl: docs/src/writing_good_rules.md

- There is only one derivative being returned, so from the fact that the user called
  `frule`/`rrule` they clearly will want to use that one.

> ## Use `ProjectTo` appropriately

Author: ```suggestion
## Ensure you remain in the primal's subspace (i.e. use `ProjectTo` appropriately)
```

Rules with abstractly-typed arguments may return incorrect answers when called with certain concrete types.

> A classic example is the matrix-matrix multiplication rule

Author: ```suggestion
A classic example is the matrix-matrix multiplication rule, an naive definition of which follows:
```

> ```

Author: ````suggestion
```julia
````

function rrule(::typeof(*), A::AbstractMatrix, B::AbstractMatrix)
    function times_pullback(ȳ)
        dA = ȳ * B'
        dB = A' * ȳ
        return NoTangent(), dA, dB
    end
    return A * B, times_pullback 
end
```
When computing `*(A, B)`, where `A isa Diagonal` and `B isa Matrix`, the output will be a `Matrix`.

> As a result, `ȳ` in the pullback will be a `Matrix`, and consequently `dA` for a `A isa Diagonal` will be a `Matrix`, which is wrong.

Author: ```suggestion
As a result, `ȳ` in the pullback will be a `Matrix`, and consequently `dA` for a `A isa Diagonal` will be a `Matrix`, which is wrong.
Note only is it the wrong type, but it can contain non-zeros off the diagonal, which is not possible, it escapes the subspace.
```

Author: Wait does this one actually have zeros on the off-diagonal? Or is `ȳ ` going to be diagonal?

Maybe a better example is `sum` which has tangent `fill(ȳ , size(A))`

Reviewer: No, it is a fully dense matrix in general, since `Diagonal` * `Matrix` is a dense `Matrix`. (Only Diagonal * Diagonal is always a Diagonal)

While a specialised rules can indeed be written for the `Diagonal` case, there are many other types and we don't want to be forced to write a rule for each of them.
Instead, `project_A = ProjectTo(A)` can be used (outside the pullback) to extract an object that knows how to project onto the type of `A` (e.g. also knows the size of the array).

> This object can be called with a tangent, `project_A(ȳ * B')`, to project it on the type `A`.

Author: ```suggestion
This object can be called with a tangent `ȳ * B'`, by doing `project_A(ȳ * B')`, to project it on the tangent space of `A`.
```

The correct rule then looks like

> ```

Author: ```suggestion
```julia
```
